### ISSUE TITLE
Servo panics exactly after 144 navigations

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#7572]
Just create this simple document:

``` html
<script>
  location.assign(location.href);
</script>
```

The page navigates 144 times (it's always 144 on my machine), and then:

```
thread 'PaintTask PipelineId(144)' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 35, message: "Resource temporarily unavailable" } }', src/libcore/result.rs:734
stack backtrace:
thread 'Constellation' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 35, message: "Resource temporarily unavailable" } }', src/libcore/result.rs:734
   1:        0x105977df0 - sys::backtrace::tracing::imp::write::hd0176e40ebe71d28Vks
   2:        0x10597aff9 - panicking::on_panic::h929e7e75bc4f2c31NXw
   3:        0x10596a192 - rt::unwind::begin_unwind_inner::h9d2d49cde66c0c4b0sw
   4:        0x10596a83d - rt::unwind::begin_unwind_fmt::h30d162ea92169f156rw
   5:        0x10597aa57 - rust_begin_unwind
   6:        0x10599fba0 - panicking::panic_fmt::ha85fd08dcab4df5d8PE
   7:        0x1049bdff8 - iter::Map<I, F>.Iterator::next::h2739781353065444127
   8:        0x1049b6049 - paint_task::WorkerThreadProxy::spawn::h09d210ec5810c6a5q3i
   9:        0x104753a00 - boxed::F.FnBox<A>::call_box::h8938302690982073553
  10:        0x104737602 - rt::unwind::try::try_fn::h7397073164000958079
  11:        0x10597aa08 - __rust_try
  12:        0x105976710 - rt::unwind::try::inner_try::h03cd26bf8e2ee1ceTow
  13:        0x1047377a7 - boxed::F.FnBox<A>::call_box::h9041483843375298480
  14:        0x105979dbd - sys::thread::Thread::new::thread_start::he62641fb2961f41aBNv
  15:     0x7fff8b2189b0 - _pthread_body
  16:     0x7fff8b21892d - _pthread_start
```

**Comment by jdm** [comment]
Probably something to do with resource exhaustion, since all of our old pipelines are never garbage collected.

**Comment by paulrouget** [comment]
It takes some more time now. And it crashes differently:

```
ERROR:servo: Thread spawn failed: Error { repr: Os { code: 35, message: "Resource temporarily unavailable" } } (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(201) }, at ../src/libcore/result.rs:788)
ERROR:servo: Thread spawn failed: Error { repr: Os { code: 35, message: "Resource temporarily unavailable" } } (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(202) }, at ../src/libcore/result.rs:788)
ERROR:servo: Thread spawn failed: Error { repr: Os { code: 35, message: "Resource temporarily unavailable" } } (thread ResourceManager, at ../src/libcore/result.rs:788)
ERROR:servo: called `Result::unwrap()` on an `Err` value: RecvError (thread LayoutWorker worker 1/3, at ../src/libcore/result.rs:788)
ERROR:servo: called `Result::unwrap()` on an `Err` value: RecvError (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(201) }, at ../src/libcore/result.rs:788)
ERROR:servo: Thread spawn failed: Error { repr: Os { code: 35, message: "Resource temporarily unavailable" } } (thread PaintThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(202) }, at ../src/libcore/result.rs:788)
ERROR:servo: called `Result::unwrap()` on an `Err` value: RecvError (thread LayoutWorker worker 1/3, at ../src/libcore/result.rs:788)
ERROR:constellation::constellation: about:failure failed
ERROR:servo: called `Result::unwrap()` on an `Err` value: RecvError (thread PaintWorker, at ../src/libcore/result.rs:788)
ERROR:servo: Thread spawn failed: Error { repr: Os { code: 35, message: "Resource temporarily unavailable" } } (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(203) }, at ../src/libcore/result.rs:788)
ERROR:servo: Thread spawn failed: Error { repr: Os { code: 35, message: "Resource temporarily unavailable" } } (thread PaintThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(203) }, at ../src/libcore/result.rs:788)
ERROR:servo: Thread spawn failed: Error { repr: Os { code: 35, message: "Resource temporarily unavailable" } } (thread Constellation, at ../src/libcore/result.rs:788)
ERROR:servo: called `Result::unwrap()` on an `Err` value: RecvError (thread LayoutWorker worker 1/3, at ../src/libcore/result.rs:788)
ERROR:servo: called `Result::unwrap()` on an `Err` value: RecvError (thread PaintWorker, at ../src/libcore/result.rs:788)
^CServo exited with return value -2
```

Would https://github.com/servo/servo/pull/11893 help?

**Comment by jdm** [comment]
I would expect #14312 to help here.

**Comment by cbrewster** [comment]
@paulrouget are you using a version after #14312 landed?

**Comment by paulrouget** [comment]
@cbrewster yes.

**Comment by paulrouget** [comment]
@asajeffrey with #14860, no crash :)

**Comment by asajeffrey** [comment]
That's good to know!

**Comment by asajeffrey** [comment]
Closed by #14860.
